client/asset/btc/btc.go

```diff
@@ -1747,336 +1733,206 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
+			btc.log.Errorf("tryRedemptionRequests getBlockHeight error: %v", err)
 			return
 		}
-		scanPoint := fmt.Sprintf("block %d", blk.Height)
-		lastScannedBlockHeight = int64(blk.Height)
-		for t := range blk.Tx {
-			tx := &blk.Tx[t]
-			found, canceled := btc.findRedemptionsInTx(scanPoint, tx, contractOutpoints)
-			totalFound += found
-			totalCanceled += canceled
-			if totalFound+totalCanceled == contractsCount {
-				break rangeBlocks
-			}
+		// If the startBlock height is lower than the lowest swap block height,
+		// ignore the startBlock.
+		if h > iHeight {
+			iHeight = h
+			iHash = startBlock
 		}
-		nextBlockHash = blk.NextHash
 	}
 
-	btc.log.Debugf("%d redemptions found, %d canceled out of %d contracts in blocks %d - %d",
-		totalFound, totalCanceled, contractsCount, startBlock.height, lastScannedBlockHeight)
-
-	// Search for redemptions in mempool if there are yet unredeemed
-	// contracts after searching this block range.
-	pendingContractsCount := contractsCount - totalFound - totalCanceled
-	if pendingContractsCount > 0 {
-		btc.findRedemptionMtx.RLock()
-		pendingContracts := make([]outPoint, 0, pendingContractsCount)
-		for _, contractOutpoint := range contractOutpoints {
-			if _, pending := btc.findRedemptionQueue[contractOutpoint]; pending {
-				pendingContracts = append(pendingContracts, contractOutpoint)
+	for iHeight <= tipHeight {
+		validReqs := make(map[outPoint]*findRedemptionReq, len(undiscovered))
+		for outPt, req := range undiscovered {
+			if iHeight >= req.blockHeight {
+				validReqs[outPt] = req
 			}
 		}
-		btc.findRedemptionMtx.RUnlock()
-		btc.findRedemptionsInMempool(pendingContracts)
-	}
-}
-
-// findRedemptionsInTx checks if any input of the passed tx spends any of the
-// specified contract outpoints. If spending info is found for any contract, the
-// contract's secret or any error encountered while trying to parse the secret
-// is returned to the redemption finder via the registered result chan; and the
-// contract is purged from the findRedemptionQueue.
-// Returns the number of redemptions found and canceled.
-func (btc *ExchangeWallet) findRedemptionsInTx(scanPoint string, tx *btcjson.TxRawResult, contractOutpoints []outPoint) (found, cancelled int) {
-	btc.findRedemptionMtx.Lock()
-	defer btc.findRedemptionMtx.Unlock()
 
-	extractSecret := func(vin int, contractHash []byte) (*chainhash.Hash, []byte, error) {
-		redeemTxHash, err := chainhash.NewHashFromStr(tx.Txid)
-		if err != nil {
-			return nil, nil, err
-		}
-		if tx.Vin[vin].ScriptSig == nil {
-			return nil, nil, fmt.Errorf("no sigScript")
-		}
-		sigScript, err := hex.DecodeString(tx.Vin[vin].ScriptSig.Hex)
-		if err != nil {
-			return nil, nil, err
-		}
-		var witness [][]byte
-		for _, hexB := range tx.Vin[vin].Witness {
-			var b []byte
-			b, err = hex.DecodeString(hexB)
-			if err != nil {
-				break
+		discovered := btc.node.searchBlockForRedemptions(validReqs, *iHash)
```

Comment:
The first time this loop runs, there may be no `undiscovered` contracts, only mempool. `len(validReqs)` would be 0 and there'd be no point to fetch this block data.